### PR TITLE
[armnn] Add caffe support with armnn @open sesame 1/16 16:11

### PR DESCRIPTION
**Pull request by kparichay**
Add caffe support with armnn
Add corresponding unit-test
Tested with CPU and GPU backend on the device

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [*]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Comment by myungjoo** [comment]
Soon, can we redesign/update APIs (single) to support caffe files with ARMNN?
Then, we will need to think about priorities of frameworks for a given model file with availability checks assisting.

**Comment by kparichay** [comment]
> Soon, can we redesign/update APIs (single) to support caffe files with ARMNN?

Yes, will be done this week.

File: ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc [hunk 1]
```diff
@@ -164,8 +165,37 @@ int
...
+      (model_path, input_map, output_vec);
+  }
+
+  /** set input/output bindings */
+  for (auto const &output_name:output_vec) {
+    outputBindingInfo.push_back (
+        parser->GetNetworkOutputBindingInfo (output_name));
+  }
+
+  for (auto const &inputs:input_map) {
```
Comment (dongju-chae) [inline ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc]: Do they have valid values when `unknown_input_dim` is true,

File: tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc [hunk 2]
```diff
@@ -352,6 +353,115 @@ TEST (nnstreamer_filter_armnn, invoke_advanced)
...
+  gst_tensors_info_init (&prop.input_meta);
+  gst_tensors_info_init (&prop.output_meta);
+  prop.output_meta.num_tensors = 1;
+  prop.output_meta.info[0].type = _NNS_FLOAT32;
+  prop.output_meta.info[0].name = g_strdup ("prob");
+  prop.input_meta.num_tensors = 1;
+  prop.input_meta.info[0].name = g_strdup ("data");
+
+  EXPECT_TRUE (g_file_get_contents (data_file, (gchar **) &input_int8.data,
+        &input_int8.size, NULL));
```
Comment (dongju-chae) [inline tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc]: uint8?

File: tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc [hunk 3]
```diff
@@ -352,6 +353,115 @@ TEST (nnstreamer_filter_armnn, invoke_advanced)
...
+      static_cast<float> (((guint8 *) input_int8.data)[idx]);
+    ((float *) input.data)[idx] -= 127.5;
+    ((float *) input.data)[idx] /= 127.5;
+  }
+
+  const GstTensorFilterFramework *sp = nnstreamer_filter_find ("armnn");
+  EXPECT_NE (sp, (void *) NULL);
+
+  output.type = _NNS_FLOAT32;
+  output.size = gst_tensor_get_element_size (input.type) * num_labels;
```
Comment (dongju-chae) [inline tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc]: output.type?

File: ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc (hunk 1 above)
Comment (kparichay) [reply ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc]: When `unknown_input_dim` is true, the dimension of the inputs are missing (`input_map.second`).
However, here we want to extract just the input names (`input_map.first`).

File: tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc (hunk 2 above)
Comment (kparichay) [reply tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc]: updated.

File: tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc (hunk 3 above)
Comment (kparichay) [reply tests/nnstreamer_filter_armnn/unittest_filter_armnn.cc]: updated.

File: ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc [hunk 4]
```diff
@@ -316,7 +353,7 @@ ArmNNCore::loadModel (const GstTensorFilterProperties * prop)
     if (status == armnn::Status::Failure)
       throw std::runtime_error ("Error loading the network.");
   }
-  catch (...) {
+  catch ( ...) {
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc]: plz fix indentation

File: ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc [hunk 5]
```diff
@@ -164,8 +165,37 @@ int
...
-  /** @todo fill this */
-  return -EPERM;
+  bool unknown_input_dim = false;
+
+  armnnCaffeParser::ICaffeParserPtr parser =
+      armnnCaffeParser::ICaffeParser::Create ();
+
+  for (auto const &inputs:input_map) {
+    if (inputs.second.GetNumDimensions () == 0) {
+      unknown_input_dim = true;
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_armnn.cc]: needs to add break?

**Pull request merged**